addons/point_of_sale/controllers/main.py

```diff
@@ -84,3 +84,64 @@ def print_sale_details(self, date_start=False, date_stop=False, **kw):
+            'address_url': f'/my/account?redirect=/pos/ticket/validate?order_reference={order_reference}%26ticket_amount={ticket_amount}',
+        }
+        # Check that an order exists for the information entered
+        pos_order = request.env['pos.order'].sudo().search([('amount_total', '=', ticket_amount), ('pos_reference', '=like', f'Order {order_reference}')])
+        if not pos_order:
+            error_message = _("No order with this reference and amount was found in the system.")
+            return request.redirect('/pos/ticket?order_reference=%s&ticket_amount=%s&error_message=%s' % (order_reference, ticket_amount, error_message))
+
+        if pos_order.account_move and pos_order.account_move.is_sale_document():  # If there is a move linked the order that is already an invoice, just return it.
+            return request.redirect('/my/invoices/%s' % pos_order.account_move.id)
+
+        # Get some basic information about the order to display in the view
+        template_vals['pos_order'] = pos_order
+
+        # Check that the user has filled its address information. Just check if a country and street has been set.
+        if not partner.country_id or not partner.street:
+            template_vals['partner_address'] = False
+        else:
+            template_vals['partner_address'] = partner._display_address(without_company=True)
+
+        template_vals['additional_requirements'] = request.env['account.move'].get_required_fields_for_country(pos_order.company_id.country_id) or False
+
+        return request.render("point_of_sale.ticket_validation_screen", template_vals)
+
+    @http.route(['/pos/ticket/make_invoice'], type='http', auth="user", website=True)
+    def print_invoice_from_order_reference(self, pos_order_id, **kwargs):
+        if not pos_order_id:  # The route is called directly
+            return request.redirect('/pos/ticket')
+        pos_order = request.env['pos.order'].sudo().browse(int(pos_order_id))
+        pos_order.partner_id = request.env.user.partner_id
+        partner_user = request.env.user.partner_id.user_ids[:1]
+        # In order for the invoices to appear on the portal and to be able to display the portal view to the user, the invoice/order company must be available to the end user.
+        if pos_order.company_id not in partner_user.company_ids:
+            partner_user.company_ids |= pos_order.company_id
+        # Get all fields related to account moves from the kwargs, to set their value as default value in the context.
+        with_context = {}
+        for key, value in kwargs.items():
+            with_context.update({f'default_{key}': value})
+        # Allowing default values for moves is important for some localizations that would need specific fields to be set on the invoice, such as Mexico.
+        pos_order.with_context(with_context).action_pos_order_invoice()
+        return request.redirect('/my/invoices/%s' % pos_order.account_move.id)
```

Comment:
⚠ more security issues here:

- as in previous method, the `sudo` is not properly verified. Any POS Order ID can be passed without any validation
- the route overwrites the customer on the order with the currently connected user, basically allowing any portal user to steal any order. That requires more validation and safety checks (e.g. the customer must not be set) + comments to explain this sensitive piece of code
- the route modifies the allowed companies of the connected user. That's absolutely out of question, it would allow any portal (and internal) user to bypass company security rules
- the route accepts arbitrary `kwargs` and puts them in the context defaults. This is not acceptable, it would allow injecting arbitrary values into invoices. The list and types of fields must be restricted.